### ISSUE TITLE
`reportUninitializedInstanceVariable` does not check that a `@dataclass` `field(init=False)`  is initialized in `post_init`

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#4701]
**Describe the bug**
Discovered by @FynnBe:  dataclasses can mark some fields with `init=False` to make those fields not appear as an `__init__` parameter. Those fields are usually then initialized in `__post_init__`.  If one forgets to initialize such a field, though, they will get no reported errors, even with `reportUninitializedInstanceVariable=error` 

**To Reproduce**
```python
# pyright: strict
# pyright: reportUninitializedInstanceVariable=error

from dataclasses import dataclass, field

@dataclass
class C:
    a: float
    b: float
    c: float = field(init=False) # no errors reported

p = C(1, 2) # no errors reported
print(p.c) # AttributeError: 'C' object has no attribute 'c'
```

**Expected behavior**
An error report at the declaration of field `c`, similar to what one would get when not initializing a normal field in a non-dataclass:

`Instance variable "c" is not initialized in __post_init__ method`

**VS Code extension or command-line**
pyright cli 1.1.295

**Comment by Tomaz-Vieira** [comment]
>  I kind of regret adding it because it's not very useful for all but a handful of code bases and coding patterns

For what it's worth, I actually find this an essential part of type checking. If you can't trust an object of type `X` to have the fields it claims to have, then what good is a type hint like `a: X`? As far as correctness goes, it's really right up there with not allowing null pointers in other languages.

It is way too easy to write wrong code without this check:

```python
# pyright: strict
# pyright: reportUninitializedInstanceVariable=none
class X:
    def blas(self):
        self.foo = 123

X().foo # checks just fine, is definitely wrong
```

**Comment by erictraut** [comment]
OK, I'm glad someone is using it. I'm guessing that you're one of just a handful.
